File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque

N,K = map(int,input().split())

board = [0 for _ in range(200010)]
visited = [0 for _ in range(200010)]
q = deque()
def bfs(x):
    q.append(x)
    visited[x] = 1
    while q:
        x = q.popleft()
        if x-1 >=0 and visited[x-1] == 0:
            board[x-1] = board[x] + 1
            q.append(x-1)
            visited[x-1]=1
        if x+1 <= 100001 and visited[x+1] == 0:
            board[x + 1] = board[x] + 1
            q.append(x+1)
            visited[x + 1] = 1
        if 2*x <= 100001 and visited[2*x] == 0:
            board[x * 2] = board[x] + 1
            q.append(2*x)
            visited[2*x] = 1
        if visited[K] == 1:
            return

bfs(N)
print(board[K])

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.util.*;
import java.io.*;

public class Main {
    public static StringBuilder sb = new StringBuilder();
    public static StringTokenizer st;
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    int MAX = 100001;
    Queue<Integer> q = new LinkedList<>();
    boolean[] visited = new boolean[200001];
    int[] time = new int[200001];

    public void sol() throws Exception {
        String[] split = br.readLine().split(" ");
        int n = Integer.parseInt(split[0]);
        int k = Integer.parseInt(split[1]);
        bfs(n, k);
        System.out.println(time[k]);

        br.close();
    }

    public void bfs(int start, int end) {
        q.add(start);
        visited[start] = true;
        time[start] = 0;
        while(!q.isEmpty()){
            int x = q.poll();
            if(x-1 >= 0 && !visited[x - 1]){
                q.add(x-1);
                visited[x - 1] = true;
                time[x-1] = time[x] + 1;
                if(x-1 == end){
                    return;
                }
            }
            if(x+1 <= MAX && !visited[x + 1]){
                q.add(x + 1);
                visited[x + 1] = true;
                time[x+1] = time[x] + 1;
                if(x-1 == end){
                    return;
                }
            }
            if (2 * x <= MAX && !visited[2 * x]) {
                q.add(2 * x);
                visited[2 * x] = true;
                time[2*x] = time[x] + 1;
                if(x-1 == end){
                    return;
                }
            }
        }
    }


    public static void main(String[] args) throws Exception {
        new Main().sol();
    }
}
